### ISSUE TITLE
regression: generic args in patterns require the turbofish syntax

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#115780]
```
[INFO] [stdout] error: generic args in patterns require the turbofish syntax
[INFO] [stdout]    --> src/lib.rs:142:16
[INFO] [stdout]     |
[INFO] [stdout] 142 |             ret<TCon> (Clone::clone(x_val), Clone::clone(y_val));
[INFO] [stdout]     |                ^
[INFO] [stdout]     |
[INFO] [stdout]    ::: src/mdo.rs:19:9
[INFO] [stdout]     |
[INFO] [stdout] 19  |         $p: pat =<< $e: expr ; $( $t: tt )*
[INFO] [stdout]     |         ------- while parsing argument for this `pat` macro fragment
[INFO] [stdout]     |
[INFO] [stdout] help: use `::<...>` instead of `<...>` to specify lifetime, type, or const arguments
[INFO] [stdout]     |
[INFO] [stdout] 142 |             ret::<TCon> (Clone::clone(x_val), Clone::clone(y_val));
[INFO] [stdout]     |                ++
[INFO] [stdout] 
[INFO] [stdout] 
```

https://crater-reports.s3.amazonaws.com/beta-1.73-1.2/beta-2023-09-10/reg/haskell_bits-0.3.0/log.txt

**Comment by fmease** [comment]
I think it just needs a `.may_recover()`

**Comment by fmease** [comment]
Minimized:

```rs
macro_rules! mdo {
    ($p: pat =<< $e: expr ; $( $t: tt )*) => {
        $e.and_then(|$p| mdo! { $( $t )* })
    };
    (ret<$ty: ty> $e: expr;) => { Some::<$ty>($e) };
}

fn main() {
    mdo! {
        x_val =<< Some(0);
        y_val =<< Some(1);
        ret<(i32, i32)> (x_val, y_val);
    };
}
```

**Comment by Nilstrieb** [comment]
cc @estebank

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23115780.20regression.3A.20generic.20args.20in.20patterns.20require.20the.20tu.E2.80.A6)).

@rustbot label -I-prioritize +P-high

**Comment by fmease** [comment]
The fix #115785 has been backported to beta in #115901. Closing.
